### PR TITLE
Ignore .numpy() under FakeTensorMode()

### LINKED ISSUE #120259: Ignore .numpy() under FakeTensorMode
### 🐛 Describe the bug

PyTorch Dynamo does not support numpy tensors, but in some cases, that might be avoidable by simply not converting the torch tensor to numpy array

```python
import numpy as np
import torch

class MyNumpyModel(torch.nn.Module):
    def __init__(self):
        super(MyNumpyModel, self).__init__()

    def forward(self, input):
        return input.numpy()

with torch._subclasses.FakeTensorMode():
    model = MyNumpyModel()
    _ = torch.export.export(model, args=(torch.randn(1000),))
```

Error

```bash
Traceback (most recent call last):
  File "/opt/pytorch/test_numpy_fake.py", line 13, in <module>
    _ = torch.export.export(model, args=(torch.randn(1000),))
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/export/__init__.py", line 186, in export
    return _export(
           ^^^^^^^^
  File "/opt/pytorch/torch/export/_trace.py", line 541, in wrapper
    raise e
  File "/opt/pytorch/torch/export/_trace.py", line 527, in wrapper
    ep = fn(*args, **kwargs)
         ^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/export/exported_program.py", line 83, in wrapper
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/export/_trace.py", line 772, in _export
    gm_torch_level = _export_to_torch_ir(
                     ^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/export/_trace.py", line 320, in _export_to_torch_ir
    gm_torch_level, _ = torch._dynamo.export(
                        ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/_dynamo/eval_frame.py", line 1314, in inner
    result_traced = opt_f(*args, **kwargs)
                    ^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/nn/modules/module.py", line 1529, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/nn/modules/module.py", line 1538, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/_dynamo/eval_frame.py", line 455, in _fn
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/nn/modules/module.py", line 1529, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/torch/nn/modules/module.py", line 1538, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/pytorch/test_numpy_fake.py", line 8, in forward
    def forward(self, input):
  File "/opt/pytorch/torch/_dynamo/utils.py", line 2020, in to_numpy_helper
    return value.numpy(force=True)
           ^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: .numpy() is not supported for tensor subclasses.
```

### Versions

main

cc @eellison @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
Fixes #120259

cc @eellison @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Review by jansel — APPROVED** [review]

File: test/test_fake_tensor.py [hunk 1]
```diff
@@ -778,6 +778,15 @@ def test__adaptive_avg_pool2d_backward(self):
             grad_in = torch.ops.aten._adaptive_avg_pool2d_backward(grad_out, inp)
             self.assertTrue(torch._prims_common.suggest_memory_format(grad_in) == torch.channels_last)
 
+    def test_export_numpy(self):
+        class MyNumpyModel(torch.nn.Module):
+            def forward(self, input):
+                return input.numpy()
```
Comment (eellison) [inline test/test_fake_tensor.py:784]: Can you add an operator which is only defined on numpy but not pytorch?

Comment (thiagocrepaldi) [reply test/test_fake_tensor.py:784]: is there such an op? :)

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your change but it seems breaking trunk on Python 3.12 https://hud.pytorch.org/pytorch/pytorch/commit/952b37145b7bb526ea5907ac574e324d274b02ee' -c weird

Trunk is very red atm, so please do a rebase to avoid hidden failures coming from broken trunk.  If this's not related to the failure in https://hud.pytorch.org/pytorch/pytorch/commit/952b37145b7bb526ea5907ac574e324d274b02ee, I'll reland it.

**Comment by thiagocrepaldi** [comment]
There seems to have 2 flaky tests remaining, but one might be relevant? See below

```bash
_______________________ FakeTensorTest.test_export_numpy _______________________
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/test/test_fake_tensor.py", line 788, in test_export_numpy
    ep = torch.export.export(MyNumpyModel(), args=(torch.randn(1000),))
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/py_3.12/lib/python3.12/site-packages/torch/export/__init__.py", line 186, in export
    return _export(
           ^^^^^^^^
  File "/opt/conda/envs/py_3.12/lib/python3.12/site-packages/torch/export/_trace.py", line 541, in wrapper
    raise e
  File "/opt/conda/envs/py_3.12/lib/python3.12/site-packages/torch/export/_trace.py", line 527, in wrapper
    ep = fn(*args, **kwargs)
         ^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/py_3.12/lib/python3.12/site-packages/torch/export/exported_program.py", line 83, in wrapper
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/py_3.12/lib/python3.12/site-packages/torch/export/_trace.py", line 772, in _export
    gm_torch_level = _export_to_torch_ir(
                     ^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/py_3.12/lib/python3.12/site-packages/torch/export/_trace.py", line 320, in _export_to_torch_ir
    gm_torch_level, _ = torch._dynamo.export(
                        ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/py_3.12/lib/python3.12/site-packages/torch/_dynamo/eval_frame.py", line 1205, in inner
    check_if_dynamo_supported()
  File "/opt/conda/envs/py_3.12/lib/python3.12/site-packages/torch/_dynamo/eval_frame.py", line 597, in check_if_dynamo_supported
    raise RuntimeError("Python 3.12+ not yet supported for torch.compile")
RuntimeError: Python 3.12+ not yet supported for torch.compile

To execute this test, run the following from the base repo dir:
     python test/test_fake_tensor.py -k test_export_numpy

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
=================================== FAILURES ===================================
_______________________ FakeTensorTest.test_export_numpy _______________________
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/test/test_fake_tensor.py", line 788, in test_export_numpy
    ep = torch.export.export(MyNumpyModel(), args=(torch.randn(1000),))
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/py_3.12/lib/python3.12/site-packages/torch/export/__init__.py", line 186, in export
    return _export(
           ^^^^^^^^
  File "/opt/conda/envs/py_3.12/lib/python3.12/site-packages/torch/export/_trace.py", line 541, in wrapper
    raise e
  File "/opt/conda/envs/py_3.12/lib/python3.12/site-packages/torch/export/_trace.py", line 527, in wrapper
    ep = fn(*args, **kwargs)
         ^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/py_3.12/lib/python3.12/site-packages/torch/export/exported_program.py", line 83, in wrapper
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/py_3.12/lib/python3.12/site-packages/torch/export/_trace.py", line 772, in _export
    gm_torch_level = _export_to_torch_ir(
                     ^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/py_3.12/lib/python3.12/site-packages/torch/export/_trace.py", line 320, in _export_to_torch_ir
    gm_torch_level, _ = torch._dynamo.export(
                        ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/py_3.12/lib/python3.12/site-packages/torch/_dynamo/eval_frame.py", line 1205, in inner
    check_if_dynamo_supported()
  File "/opt/conda/envs/py_3.12/lib/python3.12/site-packages/torch/_dynamo/eval_frame.py", line 597, in check_if_dynamo_supported
    raise RuntimeError("Python 3.12+ not yet supported for torch.compile")
RuntimeError: Python 3.12+ not yet supported for torch.compile

To execute this test, run the following from the base repo dir:
     python test/test_fake_tensor.py -k test_export_numpy

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
- generated xml file: /var/lib/jenkins/workspace/test/test-reports/python-pytest/test_fake_tensor/test_fake_tensor-[920](https://github.com/pytorch/pytorch/pull/120261/checks#step:16:921)42[938](https://github.com/pytorch/pytorch/pull/120261/checks#step:16:939)[974](https://github.com/pytorch/pytorch/pull/120261/checks#step:16:975)09d8c.xml -
=========================== short test summary info ============================
FAILED [0.0014s] test_fake_tensor.py::FakeTensorTest::test_export_numpy - RuntimeError: Python 3.12+ not yet supported for torch.compile

To execute this test, run the following from the base repo dir:
     python test/test_fake_tensor.py -k test_export_numpy
```

Maybe I should add a `@unittest.skipIf(TEST_WITH_TORCHDYNAMO, "Python 3.12+ not yet supported for torch.compile")` to the test?

**Comment by ezyang** [comment]
Yes if the test exercises Dynamo you need to prevent it from runnijg on 3.12 rn

**Pull request closed without merging**